File: langs/nodejs/forgehx/src/forgex.js
// ForgeX sponge (SPECIFICATION_X §4).

import { permute } from "./forgeperm.js";

export const STATE_WORDS = 16;
export const RATE_WORDS = 8;
export const RATE_BYTES = RATE_WORDS * 8;

const MASK64 = 0xffffffffffffffffn;

function le32(v) {
  const b = Buffer.alloc(4);
  b.writeUInt32LE(v >>> 0, 0);
  return b;
}

export class ForgeX {
  constructor() {
    this._state = new Array(STATE_WORDS).fill(0n);
    this._offset = 0;
    this._squeezing = false;
  }

  static hash(domainTag, data) {
    const x = new ForgeX();
    x.absorbDomain(domainTag, data);
    return x.squeeze(32);
  }

  static xof(domainTag, data, length) {
    if (length < 0) {
      throw new RangeError("ForgeX: negative length");
    }
    const x = new ForgeX();
    x.absorbDomain(domainTag, data);
    return x.squeeze(length);
  }

  absorbDomain(domainTag, data) {
    const tag = Buffer.from(domainTag, "ascii");
    this.absorb(le32(tag.length));
    this.absorb(tag);
    this.absorb(data);
  }

  absorb(data) {
    if (this._squeezing) {
      throw new Error("ForgeX: cannot absorb after squeezing");
    }
    const buf = Buffer.isBuffer(data) ? data : Buffer.from(data);
    let rate = this._rateBytes();
    for (let i = 0; i < buf.length; i++) {
      rate[this._offset] ^= buf[i];
      this._offset += 1;
      if (this._offset === RATE_BYTES) {
        this._writeRate(rate);
        permute(this._state);
        this._offset = 0;
        rate = this._rateBytes();
      }
    }
    this._writeRate(rate);
  }

  squeeze(length) {
    if (!this._squeezing) {
      this._padAndSwitch();
      this._squeezing = true;
    }
    const out = Buffer.alloc(length);
    let written = 0;
    let rate = this._rateBytes();
    while (written < length) {
      if (this._offset === RATE_BYTES) {
        permute(this._state);
        this._offset = 0;
        rate = this._rateBytes();
      }
      const n = Math.min(RATE_BYTES - this._offset, length - written);
      rate.copy(out, written, this._offset, this._offset + n);
      this._offset += n;
      written += n;
    }
    return out;
  }

  _padAndSwitch() {
    const rate = this._rateBytes();
    rate[this._offset] ^= 0x01;
    rate[RATE_BYTES - 1] ^= 0x80;
    this._writeRate(rate);
    permute(this._state);
    this._offset = 0;
  }

  _rateBytes() {
    const rate = Buffer.alloc(RATE_BYTES);
    for (let i = 0; i < RATE_WORDS; i++) {
      rate.writeBigUInt64LE(this._state[i] & MASK64, i * 8);
    }
    return rate;
  }

  _writeRate(rate) {
    for (let i = 0; i < RATE_WORDS; i++) {
      this._state[i] = rate.readBigUInt64LE(i * 8);
    }
  }
}

File: langs/rust/forgehx/src/forgex.rs
//! ForgeX sponge (SPECIFICATION_X §4).

use crate::forgeperm::{self, WORDS};

pub const RATE_WORDS: usize = 8;
pub const RATE_BYTES: usize = RATE_WORDS * 8;

pub struct ForgeX {
    state: [u64; WORDS],
    offset: usize,
    squeezing: bool,
}

impl ForgeX {
    pub fn new() -> Self {
        Self {
            state: [0u64; WORDS],
            offset: 0,
            squeezing: false,
        }
    }

    /// AbsorbDomain + pad + squeeze 32 bytes.
    pub fn hash(domain_tag: &str, data: &[u8]) -> [u8; 32] {
        let mut x = Self::new();
        x.absorb_domain(domain_tag, data);
        let mut out = [0u8; 32];
        x.squeeze_into(&mut out);
        out
    }

    /// AbsorbDomain + pad + squeeze `length` bytes.
    pub fn xof(domain_tag: &str, data: &[u8], length: usize) -> Vec<u8> {
        let mut x = Self::new();
        x.absorb_domain(domain_tag, data);
        let mut out = vec![0u8; length];
        x.squeeze_into(&mut out);
        out
    }

    pub fn absorb_domain(&mut self, domain_tag: &str, data: &[u8]) {
        let tag = domain_tag.as_bytes();
        self.absorb(&(tag.len() as u32).to_le_bytes());
        self.absorb(tag);
        self.absorb(data);
    }

    pub fn absorb(&mut self, data: &[u8]) {
        assert!(!self.squeezing, "ForgeX: cannot absorb after squeezing");
        let mut rate = self.rate_bytes();
        for &b in data {
            rate[self.offset] ^= b;
            self.offset += 1;
            if self.offset == RATE_BYTES {
                self.write_rate(&rate);
                forgeperm::permute(&mut self.state);
                self.offset = 0;
                rate = self.rate_bytes();
            }
        }
        self.write_rate(&rate);
    }

    pub fn squeeze_into(&mut self, out: &mut [u8]) {
        if !self.squeezing {
            self.pad_and_switch();
            self.squeezing = true;
        }
        let mut written = 0;
        let mut rate = self.rate_bytes();
        while written < out.len() {
            if self.offset == RATE_BYTES {
                forgeperm::permute(&mut self.state);
                self.offset = 0;
                rate = self.rate_bytes();
            }
            let n = (RATE_BYTES - self.offset).min(out.len() - written);
            out[written..written + n].copy_from_slice(&rate[self.offset..self.offset + n]);
            self.offset += n;
            written += n;
        }
    }

    fn pad_and_switch(&mut self) {
        let mut rate = self.rate_bytes();
        rate[self.offset] ^= 0x01;
        rate[RATE_BYTES - 1] ^= 0x80;
        self.write_rate(&rate);
        forgeperm::permute(&mut self.state);
        self.offset = 0;
    }

    fn rate_bytes(&self) -> [u8; RATE_BYTES] {
        let mut rate = [0u8; RATE_BYTES];
        for i in 0..RATE_WORDS {
            rate[i * 8..(i + 1) * 8].copy_from_slice(&self.state[i].to_le_bytes());
        }
        rate
    }

    fn write_rate(&mut self, rate: &[u8; RATE_BYTES]) {
        for i in 0..RATE_WORDS {
            let mut le = [0u8; 8];
            le.copy_from_slice(&rate[i * 8..(i + 1) * 8]);
            self.state[i] = u64::from_le_bytes(le);
        }
    }
}

impl Default for ForgeX {
    fn default() -> Self {
        Self::new()
    }
}
